File: .net/dojos/dojo1/Bowling/Bowling/Frame.cs
using System;

namespace Bowling
{
    public class Frame
    {
        public int FirstTry { get; set; }
        public int SecondTry { get; set; }

        public int Score => CalculateScore();

        public Frame Next { get; set; }

        protected virtual int CalculateScore()
        {
            if (IsStrike() && SecondTry != 0)
            {
                throw new ArgumentOutOfRangeException(nameof(SecondTry),
                    "Second try should not be play when it's strike.");
            }
            return FirstTry + SecondTry + Bonus();
        }

        private int Bonus()
        {
            if (IsStrike())
            {
                return NextBall() + NextToNextBall();
            }
            if (IsSpare())
            {
                return NextBall();
            }
            return 0;
        }

        private int NextToNextBall()
        {
            if (Next.IsStrike() && NextIsNotLast())
            {
                return Next.Next.FirstTry;
            }
            return Next.SecondTry;
        }

        private bool NextIsNotLast()
        {
            return Next.Next != null;
        }

        private int NextBall()
        {
            return Next.FirstTry;
        }

        private bool IsSpare()
        {
            return FirstTry + SecondTry == 10;
        }

        private bool IsStrike()
        {
            return FirstTry == 10;
        }
    }
}

File: dotnet/dojos/dojo1/JBG/src/main/java/bowling/sixth/Frame.java
package bowling.sixth;

public class Frame {
    private final int first;
    private final int second;
    private Frame nextFrame;
    private int ownScore;

    public Frame(int first, int second) {
        this.first = first;
        this.second = second;
        ownScore = first + second;
    }

    public int countScore() {
        return ownScore + countBonus();
    }

    private int countBonus() {
        boolean isStrike = first == 10;
        boolean isSpare = ownScore == 10;

        int bonus = 0;
        if (isStrike) {
            bonus = nextFrame.first + nextFrame.second;
        } else if (isSpare) {
            bonus = nextFrame.first;
        }
        return bonus;
    }

    public void setNext(Frame nextFrame) {

        this.nextFrame = nextFrame;
    }
}
